File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/service/BurritosToGoServiceTest.java
package com.jameskbride.eventbusmvpdemo.network.service;

import com.google.gson.Gson;
import com.jameskbride.eventbusmvpdemo.GetProfileResponseEvent;
import com.jameskbride.eventbusmvpdemo.bus.GetProfileErrorEvent;
import com.jameskbride.eventbusmvpdemo.bus.GetProfileEvent;
import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent;
import com.jameskbride.eventbusmvpdemo.bus.SecurityErrorEvent;
import com.jameskbride.eventbusmvpdemo.network.BurritosToGoApi;
import com.jameskbride.eventbusmvpdemo.network.Order;
import com.jameskbride.eventbusmvpdemo.network.ProfileResponse;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.schedulers.TestScheduler;
import okhttp3.MediaType;
import okhttp3.ResponseBody;
import retrofit2.HttpException;
import retrofit2.Response;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

public class BurritosToGoServiceTest {
    
    @Mock private BurritosToGoApi burritosToGoApi;

    private BurritosToGoService subject;

    private EventBus eventBus;
    private TestScheduler testScheduler;

    private boolean getProfileErrorEventFired;
    private GetProfileResponseEvent getProfileResponseEvent;
    private boolean networkErrorEventfired;
    private boolean securityErrorEventFired;

    @Before
    public void setUp() {
        initMocks(this);
        eventBus = EventBus.getDefault();
        testScheduler = new TestScheduler();

        subject = new BurritosToGoService(eventBus, burritosToGoApi, testScheduler, testScheduler);

        eventBus.register(this);
        subject.open();
    }

    @After
    public void tearDown() {
        eventBus.unregister(this);
        subject.close();
    }

    @Test
    public void itRegistersForGetProfileEvent() {
        ProfileResponse profileResponse = buildProfileResponseWithOrders();
        when(burritosToGoApi.getProfile("1")).thenReturn(Observable.just(profileResponse));

        eventBus.post(new GetProfileEvent("1"));
        testScheduler.triggerActions();

        assertNotNull(getProfileResponseEvent);
    }

    @Test
    public void itPostsANetworkErrorEventWhenAnIOExceptionOccurs() {
        IOException throwable = new IOException("parse exception");
        when(burritosToGoApi.getProfile("1")).thenReturn(Observable.<ProfileResponse>error(throwable));

        eventBus.post(new GetProfileEvent("1"));
        testScheduler.triggerActions();

        assertTrue(networkErrorEventfired);
    }

    @Test
    public void itPostsASecurityErrorWhenAnUnauthorizedResponseOccurs() {
        HttpException httpException = new HttpException(Response.error(401, ResponseBody.create(MediaType.parse("application/json"), "")));
        when(burritosToGoApi.getProfile("2")).thenReturn(Observable.<ProfileResponse>error(httpException));

        eventBus.post(new GetProfileEvent("2"));
        testScheduler.triggerActions();

        assertTrue(securityErrorEventFired);
    }

    @Test
    public void onGetProfileEventEmitsGetProfileErrorWhenAFailureOccurs() {
        Gson gson = new Gson();
        gson.toJson(new CustomError());
        Response<Object> throwable = Response.error(400, ResponseBody.create(MediaType.parse("application/json"), gson.toString()));
        HttpException exception = new HttpException(throwable);
        when(burritosToGoApi.getProfile("1")).thenReturn(Observable.<ProfileResponse>error(exception));

        subject.onGetProfileEvent(new GetProfileEvent("1"));
        testScheduler.triggerActions();

        assertTrue(getProfileErrorEventFired);
    }

    @Test
    public void onGetProfileEventEmitsGetProfileResponseEventWhenAResponseIsReceived() {
        ProfileResponse profileResponse = buildProfileResponseWithOrders();
        Observable<ProfileResponse> observable = Observable.just(profileResponse);
        when(burritosToGoApi.getProfile("1")).thenReturn(observable);

        subject.onGetProfileEvent(new GetProfileEvent("1"));
        testScheduler.triggerActions();

        assertEquals(profileResponse, getProfileResponseEvent.getProfileResponse());
    }

    @Subscribe
    public void onGetProfileErrorEvent(GetProfileErrorEvent getProfileErrorEvent){
        getProfileErrorEventFired = true;
    }

    @Subscribe
    public void onGetProfileResponseEvent(GetProfileResponseEvent getProfileResponseEvent) {
        this.getProfileResponseEvent = getProfileResponseEvent;
    }

    @Subscribe
    public void onNetworkErrorEvent(NetworkErrorEvent networkErrorEvent) {
        networkErrorEventfired = true;
    }

    @Subscribe
    public void onSecurityErrorEvent(SecurityErrorEvent securityErrorEvent) {
        securityErrorEventFired = true;
    }

    private ProfileResponse buildProfileResponseWithOrders() {
        List<Order> orders = new ArrayList<>();
        orders.add(new Order(1, "description"));
        ProfileResponse profileResponse = new ProfileResponse(
                "first name",
                "last name",
                "address line 1",
                "address line 2",
                "city",
                "OH",
                "12345",
                orders
        );
        return profileResponse;
    }

    private static class CustomError {

    }
}

File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/service/BurritosToGoServiceTest.kt
package com.jameskbride.eventbusmvpdemo.network.service

import com.jameskbride.eventbusmvpdemo.bus.GetProfileErrorEvent
import com.jameskbride.eventbusmvpdemo.bus.GetProfileEvent
import com.jameskbride.eventbusmvpdemo.bus.GetProfileResponseEvent
import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent
import com.jameskbride.eventbusmvpdemo.network.ProfileApi
import com.jameskbride.eventbusmvpdemo.network.Order
import com.jameskbride.eventbusmvpdemo.network.ProfileResponse
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import io.reactivex.schedulers.TestScheduler
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.MockitoAnnotations.initMocks
import java.io.IOException
import okhttp3.ResponseBody
import com.google.gson.Gson
import okhttp3.MediaType
import org.junit.Assert.*
import retrofit2.HttpException
import retrofit2.Response

class BurritosToGoServiceTest {

    @Mock private lateinit var burritosToGoApi:ProfileApi

    private lateinit var subject:ProfileService

    private lateinit var eventBus:EventBus
    private lateinit var testScheduler:TestScheduler

    private var getProfileErrorEventFired: Boolean = false
    private lateinit var getProfileResponseEvent: GetProfileResponseEvent
    private var networkErrorEventfired: Boolean = false

    @Before
    fun setUp() {
        initMocks(this)
        eventBus = EventBus.getDefault()
        testScheduler = TestScheduler()

        subject = ProfileService(eventBus, burritosToGoApi, testScheduler, testScheduler)

        eventBus.register(this)
        subject.open()
    }

    @After
    fun tearDown() {
        eventBus.unregister(this)
        subject.close()
    }

    @Test
    fun itRegistersForGetProfileEvent() {
        val profileResponse = buildProfileResponseWithOrders()
        whenever(burritosToGoApi.getProfile("1")).thenReturn(Observable.just(profileResponse))
        eventBus.post(GetProfileEvent("1"))
        testScheduler.triggerActions()

        assertNotNull(getProfileResponseEvent)
    }

    @Test
    fun onGetProfileEventEmitsGetProfileErrorWhenAFailureOccurs() {
        val gson = Gson()
        gson.toJson(CustomError())
        val throwable = Response.error<ProfileResponse>(400, ResponseBody.create(MediaType.parse("application/json"), gson.toString()))
        val exception = HttpException(throwable)
        whenever(burritosToGoApi.getProfile("1")).thenReturn(Observable.error<ProfileResponse>(exception))
        subject.onGetProfileEvent(GetProfileEvent("1"))
        testScheduler.triggerActions()

        assertTrue(getProfileErrorEventFired)
    }

    @Test
    fun onGetProfileEventEmitsGetProfileResponseEventWhenAResponseIsReceived() {
        val profileResponse = buildProfileResponseWithOrders()
        val observable = Observable.just(profileResponse)
        whenever(burritosToGoApi.getProfile("1")).thenReturn(observable)

        subject.onGetProfileEvent(GetProfileEvent("1"))
        testScheduler.triggerActions()

        assertEquals(profileResponse, getProfileResponseEvent.profileResponse)
    }

    @Test
    fun itPostsANetworkErrorEventWhenAnIOExceptionOccurs() {
        val throwable = IOException("parse exception")
        whenever(burritosToGoApi.getProfile("1")).thenReturn(Observable.error<ProfileResponse>(throwable))

        eventBus.post(GetProfileEvent("1"))
        testScheduler.triggerActions()

        assertTrue(networkErrorEventfired)
    }

    @Subscribe
    fun onGetProfileErrorEvent(getProfileErrorEvent: GetProfileErrorEvent){
        getProfileErrorEventFired = true
    }

    @Subscribe
    fun onGetProfileResponseEvent(getProfileResponseEvent: GetProfileResponseEvent) {
        this.getProfileResponseEvent = getProfileResponseEvent
    }

    @Subscribe
    fun onNetworkErrorEvent(networkErrorEvent: NetworkErrorEvent) {
        networkErrorEventfired = true
    }

    private fun buildProfileResponseWithoutOrders(): ProfileResponse {
        return ProfileResponse(
                firstName = "first name",
                lastName = "last name",
                addressLine1 = "address line 1",
                addressLine2 = "address line 2",
                city = "city",
                state = "OH",
                zipCode = "12345"
        )
    }

    private fun buildProfileResponseWithOrders(): ProfileResponse {
        return ProfileResponse(
                firstName = "first name",
                lastName = "last name",
                addressLine1 = "address line 1",
                addressLine2 = "address line 2",
                city = "city",
                state = "OH",
                zipCode = "12345",
                orderHistory = listOf(
                        Order(1, description = "description")
                )
        )
    }
}

class CustomError